gcd/maestro_gcd.go

```diff
@@ -0,0 +1,263 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gcd
+
+
+import (
+	
+	"encoding/json"
+	"os"
+	"errors"
+	"github.com/armPelionEdge/maestro/log"
+	"github.com/armPelionEdge/maestro/maestroConfig"
+	b64 "encoding/base64"
+)
+
+type gwResourceManagerRegisterArgs struct {
+	Name string `json:"name"`
+}
+
+type addResouceArgs struct {
+	Lwm2mObjects []lwm2mObject `json:"objects"`
+}
+
+type lwm2mObject struct {
+	ObjectId int 					 `json:"objectId"`
+	ObjectInstances []objectInstance `json:"objectInstances"`
+}
+
+type objectInstance struct {
+	ObjectInstanceId int `json:"objectInstanceId"`
+	Resources []resource `json:"resources"`
+}
+
+type resource struct {
+	ResourceId int  `json:"resourceId"`
+	Operations int  `json:"operations"`
+	Type string `json:"type"`
+	Value string    `json:"value"`
+}
+
+var grm_client *Client = nil   // Gateway Resource Manager Client
+var gcd_config *maestroConfig.GcdConfig // Gateway Capability Discovery config
+
+
+
+func Gcd_init(config *maestroConfig.GcdConfig) {
+
+	if(config == nil) {
+		log.MaestroErrorf("maestroGCD: No config Provided")
+		return
+	}
+	if (config.EdgeCoreSocketPath == "") {
+		log.MaestroWarnf("maestroGCD: edge_core_socketpath not provided in config file. Using default socketpath '/tmp/edge.sock'")
+		config.EdgeCoreSocketPath = "/tmp/edge.sock"
+	}
+	if (config.GatewayResources == nil) {
+		log.MaestroErrorf("maestroGCD: No gateway resources provided")
+		return
+	}
+	if(config.ConfigObjectId <= 0) { 
+		log.MaestroErrorf("maestroGCD: lwm2m objectid not provided")
+		return
+	}
+	var err error
+	gcd_config = config
+	log.MaestroInfof("maestroGCD: connecting to edge-core")
+
+	grm_client, err = grm_connect(config.EdgeCoreSocketPath)   //connect to edge-core
+	if (err == nil) {
+		defer grm_client.Close()
+	} else {
+		log.MaestroErrorf("maestroGCD: could not connect to edge-core")
+		return	
+	}
+	log.MaestroInfof("maestroGCD: successfully connected to edge-core\n")
+	objectinstanceid := 0
+	for _, gatewayResource := range config.GatewayResources {
+		err = grm_add_resource(config.ConfigObjectId, objectinstanceid, 1, 3, "string", gatewayResource.Name)
+		if (err == nil) {
+			grm_write_resource(config.ConfigObjectId, objectinstanceid, 1, 3, "string", b64.StdEncoding.EncodeToString([]byte(gatewayResource.Name)))
+		}
+		enable := "0"
+		if (gatewayResource.Enable == true) {
+			enable = "1"
+		}
+		err = grm_add_resource(config.ConfigObjectId, objectinstanceid, 2, 3, "string", string(enable))
+		if (err == nil) {
+			grm_write_resource(config.ConfigObjectId, objectinstanceid, 2, 3, "string", b64.StdEncoding.EncodeToString([]byte(enable)))
+		}
+		err = grm_add_resource(config.ConfigObjectId, objectinstanceid, 3, 3, "string", "Config")
+		if(err == nil) {
+			grm_write_resource(config.ConfigObjectId, objectinstanceid, 3, 3, "string", b64.StdEncoding.EncodeToString([]byte("Config")))
```

Comment:
@ashishwigwag Please modify this to read the content of the file and report that as default value on registration.

Reply:
@ygoyal18 Added fix, please verify.

Reply:
Verified.